### PR TITLE
Add dTrade object and related tags, mechanisms, and command

### PR DESCRIPTION
**Pull request by MusicScore**
Tested to work on 1.13.2 and 1.12.2. 1.8.8 and 1.9.4 decided to have strokes and I didn't feel like seeing if 1.10.2 and 1.11.2 could handle the stress of adding merchant trades.

### CODE REVIEW

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java [hunk 1]
```diff
@@ -736,6 +734,18 @@ public dInventory getInventory() {
                 : dInventory.mirrorBukkitInventory(getBukkitInventory()) : null;
     }
 
+    public dList getTradeRecipes() {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java]: Why is this here and not in the property

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dPlayer.java [hunk 2]
```diff
@@ -1567,6 +1564,35 @@ else if (isOnline()) {
                     .getAttribute(attribute.fulfill(1));
         }
 
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_12_R1)
+                && getPlayerEntity().getOpenInventory().getTopInventory() instanceof MerchantInventory) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/dPlayer.java]: These checks and object grabs go _inside_ the tags, not outside

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntityTrades.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        // -->
+        if (mechanism.matches("trades")) {
+            ArrayList<MerchantRecipe> recipes = new ArrayList<>();
+            for (dTrade recipe : mechanism.getValue().asType(dList.class).filter(dTrade.class)) {
+                recipes.add(recipe.getRecipe());
+            }
+            ((Merchant) entity.getBukkitEntity()).setRecipes(recipes);
+        }
+    }
+}
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntityTrades.java]: NL@EOF

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeInputs.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        // Sets the items required to make a successful trade. Use an empty input to make the trade impossible.
+        // NOTE: If a more than two items are specified, then only the first two items will be used.
+        // @tags
+        // <trade@trade.input>
+        // -->
+        if (mechanism.matches("inputs")) {
+            List<ItemStack> ingredients = new ArrayList<>();
+            dList list = mechanism.getValue().asType(dList.class);
+
+            if (!mechanism.hasValue() || list.filter(dItem.class).isEmpty()) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeInputs.java]: Don't call `list.filter(dItem.class)` repeatedly - call it once and store the result

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            ItemStack item = mechanism.getValue().asType(dItem.class).getItemStack();
+            MerchantRecipe oldRecipe = recipe.getRecipe();
+
+            MerchantRecipe newRecipe = new MerchantRecipe(item, oldRecipe.getUses(), oldRecipe.getMaxUses(), oldRecipe.hasExperienceReward());
+            newRecipe.setIngredients(oldRecipe.getIngredients());
+
+            recipe.setRecipe(newRecipe);
+        }
+    }
+}
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java]: NL@EOF

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    public String getAttribute(Attribute attribute) {
+        if (attribute == null) {
+            return null;
+        }
+
+        // <--[tag]
+        // @attribute <trade@trade.result>
+        // @returns dItem
+        // @mechanism dTradeRecipe.result
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java]: Bad meta

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeUses.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        // @description
+        // Sets the amount of times the trade has been used.
+        // @tags
+        // <trade@trade.uses>
+        // -->
+        if (mechanism.matches("uses") && mechanism.requireInteger()) {
+            recipe.getRecipe().setUses(mechanism.getValue().asInt());
+        }
+    }
+}
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeUses.java]: NL@EOF

File: plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java [hunk 8]
```diff
@@ -2685,6 +2685,46 @@ public void registerCoreMembers() {
...
+        // <e@entity.trades>
+        // <e@entity.trading_with>
+        //
+        // @Usage
+        // Use to open an unusable trade.
+        // - opentrades <trade@trade>
+        //
+        // @Usage
+        // Use to open a list of trades with an optional title.
+        // - opentrades <trade@trade[result=i@stone;inputs=li@i@stone;max_uses=9999]> "title:Useless Trades"
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java]: no `<>` if you aren't writing a tag

File: plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java [hunk 9]
```diff
@@ -2685,6 +2685,46 @@ public void registerCoreMembers() {
...
+        //
+        // @Tags
+        // <p@player.selected_trade_index>
+        // <e@entity.is_trading>
+        // <e@entity.trades>
+        // <e@entity.trading_with>
+        //
+        // @Usage
+        // Use to open an unusable trade.
+        // - opentrades <trade@trade>
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java]: no `<>` if you aren't writing a tag

File: plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/player/OpenTradesCommand.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                merchant.setRecipes(recipes);
+                player.getPlayerEntity().openMerchant(merchant, true);
+            }
+            else {
+                dB.echoError("Tried to make a nonexistent or offline player view a virtual trading inventory!");
+            }
+        }
+
+    }
+}
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/player/OpenTradesCommand.java]: NL@EOF

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java (hunk 1 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java]: Leftover from the old PR, where I had three Trade-related properties instead of only one. Will fix.

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dPlayer.java (hunk 2 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/dPlayer.java]: Will fix.

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntityTrades.java (hunk 3 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntityTrades.java]: Will add.

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeInputs.java (hunk 4 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeInputs.java]: Alrighty

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java (hunk 5 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java]: Will add.

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java (hunk 6 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeResult.java]: Oops, forgot to change that from the old (bad) PR. Will fix.

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeUses.java (hunk 7 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/properties/trade/TradeUses.java]: Will add.

File: plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java (hunk 8 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java]: This is why you don't rage-PR, kids.

Will fix.

File: plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java (hunk 9 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/BukkitCommandRegistry.java]: Will fix.

File: plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/player/OpenTradesCommand.java (hunk 10 above)
Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/scripts/commands/player/OpenTradesCommand.java]: Will add.

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dTrade.java [hunk 11]
```diff
@@ -156,12 +154,7 @@ public void applyProperty(Mechanism mechanism) {
...
-        for (Property property : PropertyParser.getProperties(this)) {
-            property.adjust(mechanism);
-            if (mechanism.fulfilled()) {
-                break;
-            }
-        }
+        CoreUtilities.autoPropertyMechanism(this, mechanism);
 
         if (!mechanism.fulfilled()) {
             mechanism.reportInvalid();
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/dTrade.java]: remove that bit^

Comment (ghost) [reply plugin/src/main/java/net/aufdemrand/denizen/objects/dTrade.java]: Done

**Comment by mcmonkey4eva** [comment]
Awesome!
Thanks!
